---
title: "Unifrac tree"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(seqinr)
library("seqRFLP")
source('~/MSK/work/microbiome_db/SQL/scripts/get_data_from_query_OTU.R')
```

```{r}
# load the table
annot <- get_data_from_query_OTU(0,'asv_annotation_blast_ag')  
seqs <- get_data_from_query_OTU(0,'asv_sequences_ag') 
```

```{r}
# should be a fasta file with 80,000 sequences 
res <- dataframe2fas(seqs %>% 
                       select(key, asv_sequence) %>% 
                       sample_n(20000), file="../output/all.fa")
```


```{bash}
# do the msa
# Fastest possible (nucleotides)
~/Downloads/SDE2/analysis/14_Day3/analysis/19_MSA/output/muscle -in ../output/all.fa -out ../output/all_msa_fast.fa -maxiters 1 -diags  
```

## Use QIIME to build the tree


