---
title: "link the shotgun fid and the actual sample and outcome table"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tidyverse)
```

## current shotgun lookup table 

```{r}
source('~/db_connect_simple.R')
connect_database(config_file = '~/dbConfig.txt')
get_table_from_database('shotgun_lookup_ad')

list_table_from_database('bile')
```


Actually the most important info with the sample: timepoint (so I know whether it is baseline) and the CR , Tox

## Italy data

```{r}
# the full italy samples and outcome joined from Antonio's two tables
italy_samp_outcome <- dt_CART_italy_samples %>% 
  full_join(dt_CART_italy_samples_and_outcome) %>% 
  distinct()

italy_samp_outcome %>% 
  write_csv('../data/shotgun_input/shotgun_italy_samp_outcome.csv')

# italy shotgun files
italy_fn <- read_csv('../../Catalog/data/full_project_11308_files.csv') %>% 
  distinct(sample_folder, .keep_all = T)

# there are 4 dups, and need to be concatenated to run
```


```{r}
samp_italy <- italy_fn %>% 
  left_join(italy_samp_outcome %>% mutate(fid = str_glue('{pid}{timepoint}')))

found1 <- samp_italy %>% 
  filter(!is.na(sampleid))

found2_df <- samp_italy %>% 
  filter(is.na(sampleid)) %>% 
  select(directory:fid) %>% 
  left_join(  italy_samp_outcome %>% 
    mutate(fid = str_replace_all(sampleid_raw, '\\+| |\\(|\\)|\\*','')  )
) %>% 
  distinct(fid, .keep_all = T) # there are two 16s samples CMT006 d172 and d168 both record as 180*

found2 <- found2_df %>% 
  filter(!is.na(sampleid))

found3_df <- found2_df %>% 
  filter(is.na(sampleid)) %>% 
  select(directory:fid) %>% 
  left_join(italy_samp_outcome %>% 
  filter(pid == 'CMT006') %>% 
  mutate(fid = str_replace_all(sampleid_raw, '\\+|\\(|\\)|\\*','')) %>% 
  filter(timepoint %in% c(90, 14))  %>% 
  mutate(fid = str_replace(fid, ' ','-'))) %>% 
  # there is one sample CMT006 that missed the datepoint that is actually day -3
  mutate(timepoint = if_else(fid == 'CMT006', -3.0, as.double(timepoint)),
         CR = if_else(fid == 'CMT006', 'Y', CR),
         Tox = if_else(fid == 'CMT006', 'N', Tox))


```

```{r}
# finally all of them matched ...

italy_pheno <- bind_rows(
  found1,
  found2, 
  found3_df
) %>% 
  select(-sample_folder) 

italy_pheno %>% 
  write_csv('../data/shotgun_pheno/italy_pheno_earlier.csv')
```

```{r}
# fix some more problems in the table including the duplicated sampleids

# now John has cleaned table and added a tab about ineligible patients 
ineligible <- readxl::read_excel('/Volumes/castoricenter/CAR-Tcell_and_Microbiota/Italy/Pt info/italy_essential_data_v4.xlsx', sheet = 2)

italy <- read_csv('../data/shotgun_pheno/italy_pheno_earlier.csv') %>% 
  # remove the row with many NA since it is a duplicate of CMT006-dm3
  filter(!is.na(sampleid)) %>% 
  # remove the ones that are ineligible
  filter(!pid %in% ineligible$pt_id) %>% 
  select(directory:fid, pid:sampleid, oligoid)

# now load the outcome for each eligible patient
eligible <- readxl::read_excel('/Volumes/castoricenter/CAR-Tcell_and_Microbiota/Italy/Pt info/italy_essential_data_v4.xlsx', sheet = 1) %>% 
  rename(pid = pt_id)

italy_all <- italy %>% 
  left_join(eligible) %>% 
  filter(!is.na(disease)) # remove the ones that have NA since they are not in the eligible patients tab either

italy_all %>% 
  write_csv('../data/shotgun_pheno/italy_all_earlier.csv')

# the sub table to be updated to shotgun lookup table
italy_up <- italy_all %>% 
  select(directory, projectID, sampleid, fid)  


```

## Upenn data

```{r}
# get the file path of the shotgun files 
upenn <- c('Project_11341')
upenn2 <- tibble(
  folder = '/Volumes/vandenBrinkLab/deep_sequencing'
)

upenn2_files <- upenn2 %>% 
  pull(folder) %>% 
  map_dfr(function(f){
    upenn %>% 
      set_names(upenn) %>% 
      imap_dfr( ~ Sys.glob(str_glue('{f}/{..1}/*/*')) %>% 
             data_frame %>% 
              mutate(..2)) %>% 
      rename(directory = names(.)[1],
             projectID = names(.)[2])
  }) %>% 
  mutate(sample_folder = str_extract(directory, 'Sample_.+$'),
         fid = str_replace(sample_folder, 'Sample_',''),
         fid = str_replace(fid, '_IGO_.+$','')) %>% 
  mutate(pid = str_replace(fid, 'MS_UPENN_','')) %>% 
  mutate(pid = str_pad(pid, 2, side = 'left', pad = '0'))

# the pheno of the ptb from John
upenn <- readxl::read_excel('/Volumes/castoricenter/CAR-Tcell_and_Microbiota/U Penn/patients/upenn_essential_data_v3.xlsx', sheet = 2) %>% 
  filter(!is.na(pt_id)) %>% 
  mutate(pt_id = as.character(pt_id)) %>% 
  mutate(pid = str_replace(pt_id, '^80',''))

upenn_samp <- readxl::read_excel('/Volumes/castoricenter/CAR-Tcell_and_Microbiota/U Penn/samples/sample_mapping_UPenn.xlsx') %>% 
  select(pt_id, OligoID) %>% 
  mutate(pt_id = as.character(pt_id))

upenn_all <- upenn2_files %>% 
  left_join(upenn) %>% 
  full_join(upenn_samp)

upenn_all %>% 
  write_csv('../data/shotgun_pheno/upenn_all_earlier.csv')

upenn_up <- upenn_all %>% 
  select(directory, projectID, OligoID, fid) %>% 
  rename(sampleid = OligoID)
```

```{r}
# 11-24 to be uploaded 
nov24_up <- bind_rows(upenn_up, italy_up)

nov24_up %>% 
  write_csv('/Volumes/vandenbrinklab/Angel_Dai/Full_human_shotgun_catalog/shotgun_cart_earlier.csv')
```


## Update the new shotgun samples to the database table 

sometimes it doesn't work you just need to restart R and run it again 

```{r}
# change the d_set_input in the upload_data_from_query_OTU_AD.R script 
source('../../../microbiome_db/SQL/scripts/upload_data_from_query_OTU_AD.R')
upload_data_from_query_OTU(1) # changed the file path of the table getting uploaded 
```

